### PR TITLE
Load sotodlib observations directly from the preproc db

### PR DESCRIPTION
**Pull request by sbiquard**
Stream observations straight from the sotodlib preprocessing database at mapmaking time, with no intermediate binary copies — removing the separate prepare-to-disk step and its disk overhead.

### Lazy preproc loading
- Generalize `AbstractLazyObservation`: `get_data` is abstract and shape inference is delegated to a new `probe_shape()` hook. Subclasses must now implement `probe_shape()`.
- Loading from a binary file moves to a new `FileBackedLazyObservation`.
- `ObservationReader` delegates structure probing to `observation.probe_shape()`, plus a refactor with no behaviour change but cleaner logic.
- Add `SOTODLibObservation.from_preproc_group` (two-layer load via `sotodlib.preproc_or_load_group` with optional downsample) and `LazyPreprocSOTODLibObservation`, which does not rely on plain binary files.

### Mapmaker changes
- Fuse the two read passes into one. Previously each observation was read twice (once in `build_model`, once in `accumulate_rhs`). Now `build_model_and_accumulate` builds the model and accumulates the hit map + RHS in a single sharded pass, so each observation is read (and, for preproc-backed observations, preprocessed) exactly once.
- Padding still balances workload across devices, but gating moved into the kernel: padding observations and failed loads are zeroed out via their masker inside the kernel (dropping out of hits, RHS, and the CG operator).
- Failed observations are collected across processes and reported in the results (`failed_observations`).
- Cache the sotodlib context per thread (built once, reused).

### CLI
- Make `furax-so-map` preproc-db-native: `--init-config`/`--proc-config` + obsids.
- Keep the legacy `.h5` path via `--obsdir`.
- Deprecate `furax-so-prepare`.
- Tolerate sotodlib preprocessing failures (a bad group no longer aborts the run); write failed vs mapped obsids in a unified file format.
- Fix outdir resolution and mapped_obsids tracking in the SO run.

### Config path handling (no more chdir on the map path)
- New `src/so_mapmaking/_preproc.py` centralizes preproc-config path knowledge in one place: the root depth, the fact that `context_file`/`archive.index` resolve against cwd while the context's own dbs resolve relative to the context file, and `{tag}` expansion.
- `furax-so-map` now **normalizes** configs (`normalize_config`) instead of `os.chdir`'ing the whole process: it absolutizes only the two cwd-sensitive keys (`context_file`, `archive.index`), leaving the context dbs untouched. This drops the process-global chdir *and* the single-root constraint — init/proc may live under different roots. Already-absolute configs (e.g. from staging) short-circuit with no copy.
- Fix a resolve-after-chdir bug in `furax-so-prepare`: config paths are now resolved before the chdir, so a relative `--init-config` no longer resolves against the wrong directory.
- New `furax-so-stage` command: copies the small sqlite indices to node-local storage (e.g. `/dev/shm`) and emits patched, fully-absolute configs, so a run reads indices locally while the large `.h5`/`.g3` data stay on the shared filesystem. Drives off the shared `_preproc` enumerator and warns on any leftover relative path. `--proc-config` is optional.

### Breaking changes
- `furax-so-prepare` is deprecated; use `furax-so-map` with `--init-config`/`--proc-config`. The legacy `--obsdir` / `.h5` path still works.
- `ObservationModel.masker` type narrowed from `AbstractLinearOperator` to `MaskOperator`; the `IdentityOperator` branch is gone.
- `pad_model` removed from `furax.mapmaking`.
- `AbstractLazyObservation` subclasses must implement the new `probe_shape()` hook.

### Tests
- Build a minimal real preproc db on the fly (obsfiledb + obsdb + preprocess archive + custom obs loader).
- Cover `probe_shape` parity, the reader `io_callback` path, and a binned map end to end.
- New `tests/so_mapmaking`: config-path resolution, `normalize_config` (cwd-independence + the absolute-config no-op), staging, and the leftover-relative warning.

### CODE REVIEW

File: src/furax/io/readers.py [hunk 1]
```diff
@@ -158,33 +175,76 @@ def _get_padding(
...
+        data, valid = jax.lax.switch(
             data_index,
-            [lambda i=i: io_callback(callback, self.out_structure, i) for i in range(self.count)],
+            [lambda i=i: io_callback(callback, result_shape, i) for i in range(self.count)],
         )
         padding = jax.lax.switch(
             data_index, [lambda i=i: self.paddings[i] for i in range(self.count)]
         )
-        return data, padding
+        return data, padding, valid
```
Comment (ArtemBasyrov) [inline src/furax/io/readers.py:240]: you output `padding` here, which is `self.padding`, but the data is padded with `_actual_padding`. Isn't it a problem?

File: src/furax/mapmaking/_model.py
Comment (ArtemBasyrov) [inline src/furax/mapmaking/_model.py:48]: and here `padding` again, but it's not used, so doesn't cause problems

File: src/furax/interfaces/sotodlib/observation.py [hunk 2]
```diff
@@ -398,3 +502,66 @@ def get_data(self, requested_fields: Collection[str] | None = None) -> SOTODLibO
...
+        self,
+        observation_id: str,
+        init_config: str | Path,
+        proc_config: str | Path | None = None,
+        detector_selection: dict[str, str] | None = None,
+        downsample: int = 1,
+        sotodlib_config: SotodlibConfig | None = None,
+    ) -> None:
+        self.observation_id = observation_id
+        # resolve now: get_data may run inside an io_callback after the process has chdir'd
```
Comment (ArtemBasyrov) [inline src/furax/interfaces/sotodlib/observation.py]: I think you removed chdir in the new commits

File: src/furax/mapmaking/mapmaker.py [hunk 3]
```diff
@@ -241,43 +250,109 @@ def make_maps(self) -> MapMakingResults:
...
+        indices = self.distribute(self.get_padded_read_indices())
+        is_real = self.distribute(self._real_observation_mask())
+        axis = jax.sharding.get_abstract_mesh().axis_names[0]
+
+        @jax.shard_map(out_specs=(P('obs'), P(), P()), check_vma=False)
+        def kernel(indices, is_real):  # type: ignore[no-untyped-def]
+            def step(carry, args):  # type: ignore[no-untyped-def]
+                hits_acc, rhs_acc = carry
+                i, real = args
+                data, padding, valid = reader.read(i)
```
Comment (ArtemBasyrov) [inline src/furax/mapmaking/mapmaker.py]: It's kinda a big comment, through several lines, but I will tag it here. Here you read the observation. And then you pad the observations earlier, you pad them with the `mode='edge'`, aka the last one. It's in `get_padded_read_indices`. But here, you read every slot, so you re-read the last observation again, only to mask it later with `_real_observation_mask`. The padded observation is loaded, preprocessed through sotodlib pipeline, only to be multiplied with False mask. It's a big waste of extra reading/pre-processing

File: src/furax/io/readers.py [hunk 4]
```diff
@@ -158,33 +175,76 @@ def _get_padding(
...
+                data = self._read_data_impure(
+                    *self.args[i], **self.keywords[i], **self.common_keywords
+                )
+            except Exception:
+                filler = self._failure_filler()
+                if filler is None:
+                    # no filler override -> crash
+                    raise
+                if i not in self.failed_indices:
+                    self.failed_indices.append(i)
```
Comment (ArtemBasyrov) [inline src/furax/io/readers.py]: I think here you might have a threading issue. When you jit and thread, two different threads can run this part here, and append the same index, because the check just before is not atomic. + two threads can mutate the same list. It doesn't cause problems now, because you just use it for a mask, but might cause problems later

File: src/furax/mapmaking/mapmaker.py
Comment (ArtemBasyrov) [inline src/furax/mapmaking/mapmaker.py:437]: There is a bug being introduced here since `LazyPreprocSOTODLibObservation.get_data` ignores requested_fields, so an observation will get a full pre-processing run fully every time it's called. The pre-processing failure here will abort the whole run, which you're trying to prevent. And this loop is also unsharded, so just to size the map, you will redo the whole pre-processing run on all processors.

File: src/furax/mapmaking/mapmaker.py [hunk 5]
```diff
@@ -241,43 +250,109 @@ def make_maps(self) -> MapMakingResults:
...
+                # RHS contribution (optionally gap-filled).
+                tod = data[ReaderField.SAMPLE_DATA]
+                if fill_gaps:
+                    if correlation_length is None or gap_filling_params is None:
+                        raise ValueError(
+                            'fill_gaps=True requires correlation_length and gap_filling_params'
+                        )
+                    gap_fill = GapFillingOperator(
+                        obs.noise_operator(correlation_length, inverse=False),  # type: ignore[arg-type]
+                        obs._get_indexer(),
```
Comment (ArtemBasyrov) [inline src/furax/mapmaking/mapmaker.py:339]: ` obs._get_indexer()` is built from the all-False mask, and is not tested in a regime with both ML solver and gap filling. The masked CG solver might experience problems here

File: src/furax/mapmaking/_reader.py [hunk 6]
```diff
@@ -154,6 +144,43 @@ def _resolve_fields(
...
+                failed.append(idx)
+                return idx, (1, 1)
+
+        local = [probe(idx) for idx in read_indices]
+        width = 1 + len(local[0][1])  # each row is (idx, *shape)
+        local_rows = np.array([(idx, *shape) for idx, shape in local], dtype=np.int64)
+
+        # Drop potential duplicates (from padding) and sort by obs index
+        all_rows = mhu.process_allgather(local_rows).reshape(-1, width)
+        shapes = [tuple(row[1:]) for row in np.unique(all_rows, axis=0)]
```
Comment (ArtemBasyrov) [inline src/furax/mapmaking/_reader.py:181]: shouldn't happen, but if the index is missing, it will corrupt the data silently. Worth adding simple `assert len(shapes) == self.count`

File: src/furax/mapmaking/mapmaker.py
Comment (sbiquard) [reply src/furax/mapmaking/mapmaker.py:437]: Good point! I don't think it's crucial to support this use case for now, but let's error when more than one process is used or if any of the observations is not file-backed.

File: src/furax/io/readers.py (hunk 1 above)
Comment (sbiquard) [reply src/furax/io/readers.py:240]: As you noted below, it's not, because it's not used downstream. But you're right. ACtually, the precomputed padding is just dead code now, let's remove it.

File: src/furax/mapmaking/mapmaker.py (hunk 3 above)
Comment (sbiquard) [reply src/furax/mapmaking/mapmaker.py]: That's right. Before, we were reading only the real observations, and then padding `ObservationModel`. But with the new fused approach we have to read all slots including dummy ones. We can get away with a "fake read" that simply fills in dummy data.

File: src/furax/mapmaking/mapmaker.py (hunk 5 above)
Comment (sbiquard) [reply src/furax/mapmaking/mapmaker.py:339]: Yes, let's also run this only for real observations that loaded successfully.

**Review by ArtemBasyrov — APPROVED** [review]
Looks good to me! My comments have been addressed, gap-filling will be handled in a separate PR
